alternator/server.cc

```diff
@@ -153,19 +153,35 @@ void server::set_routes(routes& r) {
     r.add(operation_type::POST, url("/"), handler);
 }
 
-future<> server::init(net::inet_address addr, uint16_t port) {
+future<> server::init(net::inet_address addr, std::optional<uint16_t> port, std::optional<uint16_t> tls_port, std::optional<tls::credentials_builder> creds) {
+    if (!port && !tls_port) {
+        return make_exception_future<>(std::runtime_error("Either regular port or TLS port"
+                " must be specified in order to init an alternator HTTP server instance"));
+    }
     return _executor.invoke_on_all([] (executor& e) {
-        return e.start();
-    }).then([this] {
-        return _control.start();
-    }).then([this] {
-        return _control.set_routes(std::bind(&server::set_routes, this, std::placeholders::_1));
-    }).then([this, addr, port] {
-        return _control.listen(socket_address{addr, port});
-    }).then([addr, port] {
-        slogger.info("Alternator HTTP server listening on {} port {}", addr, port);
-    }).handle_exception([addr, port] (std::exception_ptr e) {
-        slogger.warn("Failed to set up Alternator HTTP server on {} port {}: {}", addr, port, e);
+            e.start().get();
+    }).then([this, addr, port, tls_port, creds] {
+        return seastar::async([this, addr, port, tls_port, creds] {
+            if (port) {
+                _control.start().get();
+                _control.set_routes(std::bind(&server::set_routes, this, std::placeholders::_1)).get();
+                _control.listen(socket_address{addr, *port}).get();
+                slogger.info("Alternator HTTP server listening on {} port {}", addr, *port);
+            }
+            if (tls_port) {
+                _tls_control.start().get();
+                _tls_control.set_routes(std::bind(&server::set_routes, this, std::placeholders::_1)).get();
+                _tls_control.server().invoke_on_all([creds] (http_server& serv) {
+                    return serv.set_tls_credentials(creds->build_server_credentials());
+                }).get();
+
+                _tls_control.listen(socket_address{addr, *tls_port}).get();
+                slogger.info("Alternator HTTPS server listening on {} port {}", addr, *tls_port);
+            }
+        });
+    }).handle_exception([addr, tls_port, port] (std::exception_ptr e) {
+        slogger.warn("Failed to set up Alternator HTTP server on {} port {}, TLS port {}: {}",
```

Comment:
My comment now is not strictly related to this patch (which already contained this warning):
Maybe it isn't a good idea to make this a warning? If the deployer wanted to listen to listen on some port, and we couldn't, might as well just fail the server, no? So this should be an error, not a warning, and probably even throw an exception...

Reply:
Indeed, I'll make it throw instead of silently continuing the initialization.